### PR TITLE
Retry compilation of ambiguous expression using assemblies directly referenced by the current module only

### PR DESCRIPTION
**Pull request by cston**
Retry expression compilation if there are ambiguous types, retrying with a compilation that drops all assemblies that aren’t directly referenced in the current module. Retry occurs at most once. If there are errors on retry, the errors from the original compilation are reported instead. The compilation from the retry does not change the cache.

Replaces PR #1434

### CODE REVIEW

**Comment by cston** [comment]
@amcasey, @KevinH-MS, @tmat, @ManishJayaswal

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs [hunk 1]
```diff
@@ -29,11 +29,21 @@ internal override DkmCompilerId CompilerId
...
             ImmutableArray<MetadataBlock> metadataBlocks,
             Guid moduleVersionId,
-            int typeToken)
+            int typeToken,
+            bool useReferencedModulesOnly)
         {
-            var previous = appDomain.GetDataItem<MetadataContextItem<CSharpMetadataContext>>();
+            if (useReferencedModulesOnly)
+            {
+                var compilation = metadataBlocks.ToCompilationReferencedModulesOnly(moduleVersionId);
```
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: Consider adding a comment about deliberately not reading or updating the cache.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs [hunk 2]
```diff
@@ -62,17 +106,21 @@ internal static ConstantValue ReinterpretConstantValue(ConstantValue raw, Specia
...
+            missingAssemblyIdentities = ImmutableArray<AssemblyIdentity>.Empty;
+            return null;
         }
 
-        internal Diagnostic GetErrorAndMissingAssemblyIdentities(DiagnosticBag diagnostics, out ImmutableArray<AssemblyIdentity> missingAssemblyIdentities)
+        internal string GetErrorMessageAndMissingAssemblyIdentities(DiagnosticBag diagnostics, DiagnosticFormatter formatter, CultureInfo preferredUICulture, out bool useReferencedModulesOnly, out ImmutableArray<AssemblyIdentity> missingAssemblyIdentities)
         {
-            var diagnosticsEnumerable = diagnostics.AsEnumerable();
-            foreach (Diagnostic diagnostic in diagnosticsEnumerable)
+            var diagnosticsEnumerable = diagnostics.AsEnumerable().Where(d => d.Severity == DiagnosticSeverity.Error);
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs]: "errorsEnumerable"?

File: src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/EvaluationContext.vb [hunk 3]
```diff
@@ -579,6 +544,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator
             Next
         End Sub
 
+        Friend Overrides Function HasDuplicateTypesOrAssemblies(diagnostic As Diagnostic) As Boolean
+            Return False
```
Comment (amcasey) [inline src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/EvaluationContext.vb]: TODO?

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs [hunk 4]
```diff
@@ -513,6 +503,17 @@ internal CompilationContext CreateCompilationContext(CSharpSyntaxNode syntax)
             }
         }
 
+        internal override bool HasDuplicateTypesOrAssemblies(Diagnostic diagnostic)
+        {
+            switch ((ErrorCode)diagnostic.Code)
+            {
+                case ErrorCode.ERR_DuplicateImport:
```
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: There's really only one?

Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: There are others. I still need to fill out this implementation and the VB implementation and add corresponding tests.

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs [hunk 5]
```diff
@@ -3816,95 +3839,294 @@ .maxstack  1
...
+        private static void VerifyAssemblyReferences(
+            MetadataReference target,
+            ImmutableArray<MetadataReference> references,
+            ImmutableArray<AssemblyIdentity> expectedIdentities)
+        {
+            Assert.True(references.Contains(target));
+            var modules = references.SelectAsArray(r => r.ToModuleInstance(fullImage: null, symReader: null, includeLocalSignatures: false));
+            using (var runtime = new RuntimeInstance(modules))
+            {
+                var identityComparer = DesktopAssemblyIdentityComparer.Default;
```
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: Should this be tied to the comparer used in the product?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs [hunk 6]
```diff
@@ -84,9 +132,89 @@ internal Diagnostic GetErrorAndMissingAssemblyIdentities(DiagnosticBag diagnosti
...
+                formatter.Format(firstError, preferredUICulture ?? CultureInfo.CurrentUICulture);
         }
 
+        internal abstract bool HasDuplicateTypesOrAssemblies(Diagnostic diagnostic);
+
         internal abstract ImmutableArray<AssemblyIdentity> GetMissingAssemblyIdentities(Diagnostic diagnostic);
+
+        protected sealed class SimpleMessageDiagnostic : Diagnostic
+        {
+            internal readonly string Message;
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs]: Expose this through GetMessage?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs (hunk 2 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs]: Renamed 'errors'.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs [hunk 7]
```diff
@@ -84,9 +132,89 @@ internal Diagnostic GetErrorAndMissingAssemblyIdentities(DiagnosticBag diagnosti
...
+            return (simpleMessage != null) ?
+                simpleMessage.Message :
+                formatter.Format(firstError, preferredUICulture ?? CultureInfo.CurrentUICulture);
         }
 
+        internal abstract bool HasDuplicateTypesOrAssemblies(Diagnostic diagnostic);
+
         internal abstract ImmutableArray<AssemblyIdentity> GetMissingAssemblyIdentities(Diagnostic diagnostic);
+
+        protected sealed class SimpleMessageDiagnostic : Diagnostic
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs]: What does one of these look like in the debugger?  Would it be worthwhile to add a DebuggerDisplayAttribute?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionCompiler.cs [hunk 8]
```diff
@@ -264,7 +256,95 @@ private EvaluationContextBase CreateMethodContext(DkmClrInstructionAddress instr
...
+                    bool useReferencedModulesOnly;
+                    ImmutableArray<AssemblyIdentity> missingAssemblyIdentities;
+                    errorMessage = context.GetErrorMessageAndMissingAssemblyIdentities(
+                        diagnostics,
+                        this.DiagnosticFormatter,
+                        preferredUICulture: null,
+                        useReferencedModulesOnly: out useReferencedModulesOnly,
+                        missingAssemblyIdentities: out missingAssemblyIdentities);
+                    if (useReferencedModulesOnly)
+                    {
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionCompiler.cs]: Assert that there are no missing assembly identities?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 9]
```diff
@@ -20,9 +22,16 @@ internal static class MetadataUtilities
 
         /// <summary>
         /// Group module metadata into assemblies.
+        /// If 'moduleVersionId' is set, the assemblies are
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: <paramref name="moduleVersionId"/>

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 10]
```diff
@@ -76,16 +85,44 @@ internal static ImmutableArray<MetadataReference> MakeAssemblyReferences(this Im
...
+                {
+                    var referencedModules = ArrayBuilder<AssemblyIdentity>.GetInstance();
+                    referencedModules.Add(module.MetadataReader.ReadAssemblyIdentityOrThrow());
+                    referencedModules.AddRange(module.MetadataReader.GetReferencedAssembliesOrThrow());
+                    RemoveUnreferencedModules(referencesBuilder, identitiesBuilder, identityComparer, referencedModules);
+                    referencedModules.Free();
+                }
             }
 
+            if (identitiesBuilder != null)
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Merge into preceding if?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs (hunk 5 above)
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: Done.

**Comment by amcasey** [comment]
LGTM.  My only real concern is the paucity of error codes.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs [hunk 11]
```diff
@@ -81,13 +78,20 @@ internal sealed class EvaluationContext : EvaluationContextBase
                 previous.Compilation :
                 metadataBlocks.ToCompilation();
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: Consider:  Would it be simpler to keep just one overload of CreateTypeContext, remove the previous parameter and pull this check out into the caller?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs [hunk 12]
```diff
@@ -44,19 +45,41 @@ static void M()
...
+            var context = EvaluationContext.CreateMethodContext(
+                default(CSharpMetadataContext),
+                blocks,
+                symReader,
+                moduleVersionId,
+                methodToken: methodToken,
+                methodVersion: 1,
+                ilOffset: ilOffset,
+                localSignatureToken: localSignatureToken);
+
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: Can the above code be extracted as a helper (similar to the way it was before).  It *seems* like we're repeating this same pattern in subsequent tests.

**Comment by KevinH-MS** [comment]
:+1:

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs (hunk 1 above)
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: Added.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs (hunk 11 above)
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: I agree, ideally we'd have a single CreateTypeContext without 'previous'. For now though, this allows testing the re-use and it matches the pair of methods for CreateMethodContext. Let's consider splitting this out later.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs (hunk 4 above)
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: There are others. I've added a few more, here and in VB but the set is not exhaustive yet.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs (hunk 11 above)
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: Ok

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs (hunk 12 above)
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: It's difficult to find the right set of helper methods, without generating many similar overloads. Will leave as is for now, with fewer helpers and more test code calling directly into the EC.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionCompiler.cs (hunk 8 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionCompiler.cs]: Done.

**Comment by cston** [comment]
Fixes #1264

**Pull request merged**